src/Arcus.Security.Tests.Unit/KeyVault/KeyVaultSecretProviderTests.cs

```diff
@@ -25,38 +31,71 @@ public void KeyVaultSecretProvider_CreateWithHttpScheme_ShouldFailWithUriFormatE
             string uri = $"http://{Guid.NewGuid():N}.vault.azure.net/";
 
             // Act & Assert
-            Assert.ThrowsAny<UriFormatException>(() => new KeyVaultSecretProvider(null, new KeyVaultConfiguration(uri)));
+            Assert.ThrowsAny<UriFormatException>(() => new KeyVaultSecretProvider((IKeyVaultAuthentication) null, new KeyVaultConfiguration(uri)));
         }
 
         [Fact]
         public void KeyVaultSecretProvider_CreateWithoutUri_ShouldFailWithArgumentException()
         {
-            // Arrange
-            string uri = null;
-
             // Act & Assert
-            Assert.ThrowsAny<ArgumentException>(() => new KeyVaultSecretProvider(new AzureKeyVaultAuthenticatorDummy(), new KeyVaultConfiguration(uri)));
+            Assert.ThrowsAny<ArgumentException>(
+                () => new KeyVaultSecretProvider(
+                    new AzureKeyVaultAuthenticatorDummy(), 
+                    new KeyVaultConfiguration(rawVaultUri: null)));
         }
 
         [Fact]
         public void KeyVaultSecretProvider_CreateWithoutClientFactory_ShouldFailWithArgumentException()
         {
             // Arrange
-            string uri = $"https://{Guid.NewGuid():N}.vault.azure.net/";
+            string uri = GenerateVaultUri();
 
             // Act & Assert
-            Assert.ThrowsAny<ArgumentException>(() => new KeyVaultSecretProvider(null, new KeyVaultConfiguration(uri)));
+            Assert.ThrowsAny<ArgumentException>(() => new KeyVaultSecretProvider((IKeyVaultAuthentication) null, new KeyVaultConfiguration(uri)));
         }
 
         [Fact]
         public void KeyVaultSecretProvider_CreateWithValidArguments_ShouldSucceed()
         {
             // Arrange
-            string uri = $"https://{Guid.NewGuid():N}.vault.azure.net/";
+            string uri = GenerateVaultUri();
 
             // Act & Assert
             var secretProvider = new KeyVaultSecretProvider(new AzureKeyVaultAuthenticatorDummy(), new KeyVaultConfiguration(uri));
             Assert.NotNull(secretProvider);
         }
+
+        [Fact]
+        public async Task KeyVaultSecretProvider_GetsSecretValue_AfterRetriedTooManyRequestException()
+        {
+            // Arrange
+            string secretName = $"secret-name-{Guid.NewGuid()}";
+            string expected = $"secret-value-{Guid.NewGuid()}";
+
+            var keyVaultClient = new SimulatedKeyVaultClient(
+                request => throw new KeyVaultErrorException("Sabotage secret retrieval with TooManyRequests")
+                {
+                    Response = new HttpResponseMessageWrapper(
+                        new HttpResponseMessage(HttpStatusCode.TooManyRequests), 
+                        "some HTTP response content to ignore")
+                },
+                request => new SecretBundle(value: expected));
```

Comment:
Since we don't need request here, can't we just take in a bundle?

Reply:
Yes, but I thought if we would like to black-box test indirect outputs from that we send via the key secret provider; we would have the same values as in the simulated key vault client.
Seems to me that this makes it easier to understand what the simulated client does.

Reply:
Sure, but do we actually need it now? If not, I would keep it simple - The YAGNI way

Reply:
Ah, yes. OK